prov/util/src/util_buf.c

```diff
@@ -48,6 +48,46 @@ static inline void util_buf_set_ftr(union util_buf *buf,
 	*buf_ftr = *ftr;
 }
 
+static void util_buf_region_free_mem(struct util_buf_pool *pool,
+				     struct util_buf_region *buf_region)
+{
+	int ret;
+
+	if (pool->attr.free_hndlr)
+		pool->attr.free_hndlr(pool->attr.ctx, buf_region->context);
+	if (pool->attr.is_mmap_region) {
+		ret = ofi_free_hugepage_buf(buf_region->mem_region,
+					    buf_region->size);
+		if (ret) {
+			FI_DBG(&core_prov, FI_LOG_CORE,
+			       "Huge page free failed: %s\n",
+			fi_strerror(-ret));
+			assert(0);
+		}
+	} else {
+		ofi_freealign(buf_region->mem_region);
+	}
+}
+
+void util_buf_region_cleanup(struct util_buf_pool *pool,
+			     struct util_buf_region *buf_region)
+{
+	size_t i;
+	union util_buf *util_buf;
+
+	dlist_remove(&buf_region->entry);
+	dlist_insert_tail(&buf_region->entry, &pool->free_region_list);
```

Comment:
@dmitrygladkov  I mean here. Why do we need to store the buf_region in a list when we are freeing the memory block associated with it?

Reply:
This is needed to not allocate new buf_region and retrieve from this list already created and only allocate memory for buffers.

Reply:
We would anyway have to allocate memory for buffers which I think would be much higher size than util_buf_region. So this optimization may not be that important. What do you think?